src/chunk.c

```diff
@@ -4665,6 +4665,116 @@ add_foreign_table_as_chunk(Oid relid, Hypertable *parent_ht)
 	chunk_add_inheritance(chunk, parent_ht);
 }
 
+void
+ts_chunk_merge_on_dimension(Chunk *chunk, const Chunk *merge_chunk, int32 dimension_id)
+{
+	const DimensionSlice *slice, *merge_slice;
+	int num_ccs, i;
+	bool dimension_slice_found = false;
+
+	if (chunk->hypertable_relid != merge_chunk->hypertable_relid)
+		ereport(ERROR,
+				(errmsg("cannot merge chunks from different hypertables"),
+				 errhint("chunk 1: \"%s\", chunk 2: \"%s\"",
+						 get_rel_name(chunk->table_id),
+						 get_rel_name(merge_chunk->table_id))));
+
+	for (int i = 0; i < chunk->cube->num_slices; i++)
+	{
+		if (chunk->cube->slices[i]->fd.dimension_id == dimension_id)
+		{
+			slice = chunk->cube->slices[i];
+			merge_slice = merge_chunk->cube->slices[i];
+			dimension_slice_found = true;
+		}
+		else if (chunk->cube->slices[i]->fd.id != merge_chunk->cube->slices[i]->fd.id)
+		{
+			/* If the slices do not match (except on time dimension), we cannot merge the chunks. */
+			ereport(ERROR,
+					(errmsg("cannot merge chunks with different partitioning schemas"),
+					 errhint("chunk 1: \"%s\", chunk 2: \"%s\" have different slices on "
+							 "dimension ID %d",
+							 get_rel_name(chunk->table_id),
+							 get_rel_name(merge_chunk->table_id),
+							 chunk->cube->slices[i]->fd.dimension_id)));
+		}
+	}
+
+	if (!dimension_slice_found)
+		ereport(ERROR,
+				(errmsg("cannot find slice for merging dimension"),
+				 errhint("chunk 1: \"%s\", chunk 2: \"%s\", dimension ID %d",
+						 get_rel_name(chunk->table_id),
+						 get_rel_name(merge_chunk->table_id),
+						 dimension_id)));
+
+	if (slice->fd.range_end != merge_slice->fd.range_start)
+		ereport(ERROR,
+				(errmsg("cannot merge non-adjacent chunks over supplied dimension"),
+				 errhint("chunk 1: \"%s\", chunk 2: \"%s\", dimension ID %d",
+						 get_rel_name(chunk->table_id),
+						 get_rel_name(merge_chunk->table_id),
+						 dimension_id)));
+
+	num_ccs =
+		ts_chunk_constraint_scan_by_dimension_slice_id(slice->fd.id, NULL, CurrentMemoryContext);
+
+	if (num_ccs <= 0)
+		ereport(ERROR,
+				(errmsg("missing chunk constraint for dimension slice"),
+				 errhint("chunk: \"%s\", slice ID %d",
+						 get_rel_name(chunk->table_id),
+						 slice->fd.id)));
```

Comment:
Definitely not a normal case, will add a comment like suggested.

Reply:
Looks good.

We have added a macro `Ensure` that can be used for situations like this. It behaves as an `Assert` in debug builds so that you get a segfault and can investigate the issue, but print an error in release builds so that you can capture the case without a server crash in the wild.